vernac/comInductive.ml

```diff
@@ -145,16 +126,50 @@ let pretype_ind_arity env sigma (loc, c, impls, pseudo_poly) =
     in
     sigma, (t, Retyping.relevance_of_sort s, concl, impls)
 
-let interp_cstrs env sigma impls mldata arity ind =
+(* ind_rel is the Rel for this inductive in the context without params.
+   n is how many arguments there are in the constructor. *)
+let model_conclusion env sigma ind_rel params n arity_indices =
+  let model_head = EConstr.mkRel (n + Context.Rel.length params + ind_rel) in
+  let model_params = Context.Rel.to_extended_vect EConstr.mkRel n params in
+  let sigma,model_indices =
+    List.fold_right
+      (fun (_,t) (sigma, subst) ->
+        let t = EConstr.Vars.substl subst (EConstr.Vars.liftn n (List.length subst + 1) (EConstr.Vars.liftn 1 (List.length params + List.length subst + 1) t)) in
+        let sigma, c = Evarutil.new_evar env sigma t in
+        sigma, c::subst)
+      arity_indices (sigma, []) in
+  sigma, EConstr.mkApp (EConstr.mkApp (model_head, model_params), Array.of_list (List.rev model_indices))
+
+let interp_cstrs env (sigma, ind_rel) impls params ind arity =
   let cnames,ctyps = List.split ind.ind_lc in
-  (* Complete conclusions of constructor types if given in ML-style syntax *)
-  let ctyps' = List.map2 (complete_conclusion mldata) cnames ctyps in
+  let arity_indices, cstr_sort = Reductionops.splay_arity env sigma arity in
   (* Interpret the constructor types *)
-  let sigma, (ctyps'', cimpls) =
+  let interp_cstr sigma ctyp =
+    let flags =
+      Pretyping.{ all_no_fail_flags with
+                  use_typeclasses = false;
```

Comment:
This solves the compcert issue. Typeclasses are still resolved at the end of the typechecking for the mutual block. Waiting until then may change which error is shown when a declaration has multiple issues but should otherwise be fine.